src/FtpClient.php

```diff
@@ -485,20 +485,21 @@ public function removeDirectory(string $path) : void
     /**
      * Returns the size of the given file.
      *
-     * @param string $path The remote file path.
+     * @param string $remoteFile The remote file.
      *
      * @return int The file size.
      *
      * @throws FtpException If not connected, or an error occurs.
      */
-    public function getSize(string $path) : int
+    public function getFileSize(string $remoteFile) : int
     {
         $this->throwIfNotConnected();
+        $this->throwIfNotExists($remoteFile);
```

Comment:
I'm not sure if I want to perform 2 FTP operations in a call such as `getSize()`, as it makes the operation non-atomic. What's the current behaviour if we call `ftp_size()` on a dir? Does it error or does it return some value?

Reply:
If we use the `ftp_size ` on a directory or a nonexisting file the method returns the value `-1`, in this situation I recommend to notify the user in every possible situation. 

* Throwing an exception if giving file is a directory.
* Throwing an exception if the giving file in not exists on the server.

From a performance perspective, I can say that the `$this->throwIfNotExists($remoteFile)` can take a considerable time (almost twice), but the benefit is improving the user experience.

Reply:
I agree, but as I said I'd love to make the operations atomic as far as possible: if we perform 2 FTP calls, something may have happened by the time the second command is executed.

If `ftp_size` returns -1 for directories AND non-existent files (does it?), maybe we can just check for this value and throw an exception?

Reply:
I also like this approach to make functions atomic as possible, but for this situation, I think it's a personal preference. If you want to give the user a clear error message on why exactly the method was failing, or you can just throw a generic exception for both situations, For me, I prefer the first approach.

I don't really understand the:

> if we perform 2 FTP calls, something may have happened by the time the second command is executed.

The two FTP calls will execute synchronously right? what you mean by this!

Reply:
I mean that a concurrent FTP session, or any other program on the server may have deleted the file between your "exists" and your "get_size" commands. So while correct 99% of the time, your useful exception message is not guaranteed to be relevant anyway. I'm not saying that atomicity is a goal that can be achieved with FTP, but if we cannot guarantee it, we have to be very explicit about it (in docs, and in exception messages).

Reply:
This makes sense, but the chance for that happening is really really small especially in a fast connection, for me I'm not going to worry so much about that, and as I said the is a personal preference, for the [Lazzard/FtpClient](https://github.com/lazzard/php-ftp-client) client methods I'm always doing similar checks and throwing exceptions if something doesn't make sense.
So I'll remove the `throwIfNotExists` method and keep the `getFileSize` name, right?

Reply:
Now that I think about it, the tested conditions are not enough: you don't handle other errors such as permission errors. If you encounter a permission issue, your code would currently say "xx is a directory"...

I'm not sure if it's worth it to try and have a nice error message for every single error condition, especially if the end goal is not to have specific exception classes like `FileExistsException` (which is not a goal for this library right now). Maybe just broaden the exception message with "either the file does not exist, is not readable, or is a directory"?

> So I'll remove the throwIfNotExists method and keep the getFileSize name, right?

Yes, go ahead.

Reply:
> Now that I think about it, the tested conditions are not enough: you don't handle other errors such as permission errors. If you encounter a permission issue, your code would currently say "xx is a directory"...

Actually, with this method, we do not have a permission issue as long as the method doesn't do one of write/read/execute operations on the server, you can test the method with a remote file with only an execute `---x--x--x` file permission, and you can still get the file size. 

> I'm not sure if it's worth it to try and have a nice error message for every single error condition, especially if the end goal is not to have specific exception classes like FileExistsException (which is not a goal for this library right now). Maybe just broaden the exception message with "either the file does not exist, is not readable, or is a directory"?

For this library purposes, I think throwing an `Unable to get the size of file $path` is fine. what do you think?

Reply:
You mean, change nothing? ;-)

Reply:
If we change the exception message to `either the file does not exist, is not readable, or is a directory` then we have to change the other exception messages of other methods like `rename`, `delete`, `listDirectory`,  and as you said you don't like to have a specific exception message for every single situation, so yeah I see the to keep the exception message as it was, and just changing the method name to `getFileSize`.